irohad/model/converters/impl/json_command_factory.cpp

```diff
@@ -491,49 +425,36 @@ namespace iroha {
         return document;
       }
 
-      std::shared_ptr<Command> JsonCommandFactory::deserializeTransferAsset(
-          const Document &command) {
-        auto transfer_asset = std::make_shared<TransferAsset>();
-
-        // src_account_id
-        transfer_asset->src_account_id = command["src_account_id"].GetString();
-
-        // dest_account_id
-        transfer_asset->dest_account_id =
-            command["dest_account_id"].GetString();
-
-        // asset_id
-        transfer_asset->asset_id = command["asset_id"].GetString();
-
-        // amount
-        auto json_amount = command["amount"].GetObject();
-        transfer_asset->amount.int_part = json_amount["int_part"].GetUint64();
-        transfer_asset->amount.frac_part = json_amount["frac_part"].GetUint64();
-
-        return transfer_asset;
+      optional_ptr<Command> JsonCommandFactory::deserializeTransferAsset(
+          const Value &document) {
+        auto des = makeFieldDeserializer(document);
+        return make_optional_ptr<TransferAsset>() |
+               des.String(&TransferAsset::src_account_id, "src_account_id") |
+               des.String(&TransferAsset::dest_account_id, "dest_account_id") |
+               des.String(&TransferAsset::asset_id, "asset_id") |
+               des.Object(&TransferAsset::amount, "amount") |
+               transform;
       }
 
       // Abstract
       Document JsonCommandFactory::serializeAbstractCommand(
           std::shared_ptr<Command> command) {
-        auto it = serializers_.find(typeid(*command));
-        if (it != serializers_.end()) {
-          return (this->*it->second)(command);
-        }
-        return Document();
+        return (this->*serializers_.at(typeid(*command)))(command);
```

Comment:
it does not look like a verbose way to get and execute handler from the handler map, simplify please

Reply:
also `std::vector<T>::at` throws if given key is not found, we should handle this somewhere

Reply:
Maybe something like this?
```
std::unordered_map<std::type_index, Serializer> serializers_;
std::unordered_map<std::string, Deserializer> deserializers_;

template <typename T, typename MAP>
nonstd::optional<T> handler_exec(const MAP& map, auto &&key, auto &&arg){
	auto it = map.find(std::forward<decltype(key)>(key));
	if(it != map.end()) 
		return (*it)(std::forward<decltype(arg)>(arg));
}
// slightly better but still bad 
handler_exec(this->serializers_, typeid(*command), command);
handler_exec(this->deserializers_, command_type, document);
```

Reply:
>throws if given key is not found, we should handle this somewhere

This method is for serialization, so we can ensure that only valid commands are passed to this method.